File: Singleton/Singleton.py
from __future__ import annotations

class Logger:
    _logs = []
    _instances = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            instance = super().__call__(*args, **kwargs)
            cls._instances[cls] = instance
        return cls._instances[cls]

    def debug(self, message) -> None:
        self._logs.append(message)
        print(f"[DEBUG]: {message}")

    def get_logs(self):
        return self._logs


class FirstLogger:
    _logger: Logger

    def __init__(self) -> None:
        self._logger = Logger()

    def display_message(self) -> None:
        self._logger.debug("First implementations")


class SecondLogger:
    _logger: Logger

    def __init__(self) -> None:
        self._logger = Logger()

    def display_message(self) -> None:
        self._logger.debug("Second implementations")


firstLoggerInstance = FirstLogger()
firstLoggerInstance.display_message()

secondLoggerInstance = SecondLogger()
secondLoggerInstance.display_message()


logger = Logger()
print(logger.get_logs())

File: Singleton/SIngleton.ts
/**
 * Singleton logger
 */
class Logger {
  private logs: any[];
  private static instance: Logger;

  private constructor() {}

  public static getInstance(): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }

    return Logger.instance;
  }

  public debug(message: any) {
    this.logs.push(message);
  }

  public getLogs(): any[] {
    return this.logs;
  }
}

// Implementation of Logger class.
class FirstLogger {
  private logger: Logger;

  constructor() {
    this.logger = Logger.getInstance();
  }

  public displayMessage() {
    this.logger.debug('Hello world from first logger.');
  }
}

// Implementation of Logger class.
class SecondLogger {
  private logger: Logger;

  constructor() {
    this.logger = Logger.getInstance();
  }

  public displayMessage() {
    this.logger.debug('Hello world from second logger.');
  }
}

let firstLoggerInstance = new FirstLogger();
firstLoggerInstance.displayMessage();

let secondLoggerInstance = new SecondLogger();
secondLoggerInstance.displayMessage();

let logger = Logger.getInstance();
console.log(logger.getLogs());
